docs/rfc/rfc-8.adoc

```diff
@@ -0,0 +1,410 @@
+link:https://arbiscan.io/token/0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18[`0xc96f2715e2a242d50d1b0bc923dbe1740b8ecf18`]
+which is not the canonical USDC on Arbitrum. The canonical USDC on Arbitrum is
+link:https://arbiscan.io/token/0xff970a61a04b1ca14834a43f5de4533ebddb5cc8[`0xff970a61a04b1ca14834a43f5de4533ebddb5cc8`].
+
+Wormhole cannot magically mint all tokens on Arbitrum or other L2s. Wormhole
+mints its own token representing the bridged asset that may be swapped to the
+canonical representation on the given L2 assuming there is a market for such
+a swap. The overview of liquid markets on each target chain is available in the
+link:https://docs.wormhole.com/wormhole/overview-liquid-markets[Wormhole documentation].
+Not every chain and token pair has a market. 
+
+To fully use the capabilities of the Wormhole portal and make the user
+experience seamless, we will implement a gateway contract having the authority to
+wrap Wormhole-specific tokens into TBTC tokens in a 1:1 ratio. This contract
+will have a minting authority for TBTC on L2. This way, no liquid market has to
+exist on any target chain for users to be able to cross the Wormhole bridge into
+the canonical TBTC.
+
+IMPORTANT: The requirement for a seamless user experience is that the Wormhole
+team has to integrate the step of wrapping the token using the gateway contract
+into the Wormhole bridge flow, in the same transaction as the L2 confirmation.
+
+=== Multisig governance
+
+The Governance in the initial implementation should be based on Gnosis Safe
+6-of-9 Threshold Council Multisig. Since the Gnosis Safe for Threshold Council
+was deployed using Safe Proxy factory v1.3.0, it
+link:https://help.safe.global/en/articles/5267779-i-sent-assets-to-a-safe-address-on-the-wrong-network-any-chance-to-recover[should be possible]
+to replay the same transaction creating Gnosis Safe with the same address on the
+supported L2s and sidechains.
+
+The Governance will be able to add and remove minters to L2/sidechain TBTC
+canonical contract.
+
+=== Code organization
+
+The code that exists in the `solidity` directory should contain components
+specific to L1 and generic L2 components that will be reused between EVM
+L2 and sidechain implementations. The code specific to individual chains should
+be placed in a chain-specific directory, in a separate NPM project:
+`cross-chain/{$chainName}`.
+
+For example:
+
+* `cross-chain/arbitrum`,
+* `cross-chain/polygon`,
+* `cross-chain/optimism`.
+
+Each `cross-chain` project should contain L1 and L2
+contracts specific to that individual subchain. This separation will allow us to
+abstract out the complexity of deployment and avoid redeploying all L1 testnet
+contracts in case a single change in one of L2 contracts is needed.
+
+This organization of the code will also allow us to not add subchain-specific
+dependencies to the L1 project and to deploy NPM packages separately:
+
+* `@keep-network/tbtc-v2-arbitrum`,
```

Comment:
Are the multiple NPM packages overkill?

Reply:
Having it all in one `solidity/` project and one `@keep-network/tbtc-v2` NPM package was my initial take but then I realized there are some complications.

First, if you `npm i @keep-network/tbtc-v2` and go to `node_modules/@keep-network/tbtc-v2/artifacts` and browse JSON artifacts, you will notice there is nothing about chain ID in JSON. Some contracts from the NPM package would be deployed on Ethereum, some others on Optimism, some others on Polygon, and some others on Arbitrum with no sign in the artifact JSON where the contract comes from. There is a potential workaround: we could reflect it in the directory structure, e.g. `artifacts/arbitrum` would be from Arbitrum, `artifacts/optimism` would be from Optimism. Though, currently, the deployer plugin puts all artifacts in a single directory with no nesting. I am not sure if this is easy to change.

Second, there are some dependencies that are L2-specific. Optimism has its modified Solidity compiler, once we start integrating native bridges there will be more. They may require a specific version of node, for example. This is already a pain for tbtc-v2 and bcrypto, see #509, #518.

Finally, I think keeping them in separate packages will make the deployment and development easier. There are already quite a few deployment steps to deploy tbtc-v2 to testnet (see [`solidity/deploy`](https://github.com/keep-network/tbtc-v2/tree/e740d15932c8b9cedc55d335538b5d197c25ede2/solidity/deploy)). If we do it all in one package, we'll have to add L2 deployment to the checklist and then go back to L1 to wire up gateways. Problems on one L2 will fail the entire deployment and NPM package publication. If we keep them separate, we do not have to repeat all the heavy L1 deployment if we need to apply a small fix to a gateway on one L2.

tl;dr; I think we keep the risk under control this way.